### PR TITLE
Bug 1991662: Catalog switcher

### PR DESCRIPTION
**Pull request by kevinrizza**
* Implement catalog switching proposal

- implements catalog switching proposal
See https://github.com/operator-framework/enhancements/blob/master/enhancements/automatic-catalog-switching-annotations.md
- picks up v0.10.3
- change sed command in copy_crds.sh to be posix compliant

The command + in sed is non-portable. The posix compliant version of
this command is {1,}

This change allows this script to run on a mac too (which uses BSD sed)

See https://riptutorial.com/sed/topic/9436/bsd-macos-sed-vs--gnu-sed-vs--the-posix-sed-specification

Note: GVK was implemented but disabled due to security concerns.
Future versions will address these concerns

Signed-off-by: John Hunkins <jhunkins@us.ibm.com>

* Address review comments from ecordell

- reorganize imports
- change logging common case to debug

Signed-off-by: John Hunkins <jhunkins@us.ibm.com>

* removed all watcher related code

- this is the commit that should be reviewed later
when it comes time to re-enable the GVK code

Signed-off-by: John Hunkins <jhunkins@us.ibm.com>

* update vendor to remove dynamic lister code

Signed-off-by: John Hunkins <jhunkins@us.ibm.com>

* fix incorrect names and improper error usage

- fix names that don't match Golang convention
- fix misspelled package name
- return nil for sync function casting

Signed-off-by: John Hunkins <jhunkins@us.ibm.com>

* change code block organization and add mutex

Signed-off-by: John Hunkins <jhunkins@us.ibm.com>

* remove all traces of GVK template feature

- clean up every instance of the GVK templating feature
- add last minute change for RWMutex instead of Mutex

- this is the second commit that should be reviewed later
when it comes time to re-enable the GVK code

Signed-off-by: John Hunkins <jhunkins@us.ibm.com>

* return status update errors and adjust logging

Signed-off-by: John Hunkins <jhunkins@us.ibm.com>

* address review comments

- update logging for clarity
- clean up code blocks
- refactor catalogsource helper functions:
  - remove mutex
  - remove two step get/update* in preference to update* even though
    this may result in more "the object has been modified" errors
  - rename helper functions to clarify usage
  - clarify the update helper function docs

Signed-off-by: John Hunkins <jhunkins@us.ibm.com>

Upstream-repository: operator-lifecycle-manager
Upstream-commit: 5f6fcd14f08a7e2a81b016fa83132ea68242f335

### CODE REVIEW

**Comment by timflannagan** [comment]
This lgtm to me but we're waiting for either exception approval or this to land as a bug in z-stream.

/hold

**Comment by kevinrizza** [comment]
```
 Status: "Failure",
            Message: "Operation cannot be fulfilled on catalogsources.operators.coreos.com \"catalog-5dh25\": the object has been modified; please apply your changes to the latest version and try again",
            Reason: "Conflict",
            Details: {
                Name: "catalog-5dh25",
                Group: "operators.coreos.com",
                Kind: "catalogsources",
                UID: "",
                Causes: nil,
                RetryAfterSeconds: 0,
            },
            Code: 409,
```

Looks like we're not resilient to cache invalidation and need to add retry logic. I believe these failures are valid

/hold

so we don't keep wasting test resources

**Comment by kevinrizza** [comment]
/hold cancel

tests updated, should avoid that flake now

File: staging/operator-lifecycle-manager/pkg/lib/catalogsource/catalogsource_update.go
Comment (dinhxuanvu) [inline staging/operator-lifecycle-manager/pkg/lib/catalogsource/catalogsource_update.go:94]: There is a bug in this function as this `Update` call only updates the spec, not the status. Another call `UpdateStatus` is needed here.
This is the downstream PR so this needs to be fixed upstream first.

**Review by dinhxuanvu — APPROVED** [review]
/lgtm

**Pull request merged**
